Fix missing ActiveRecord connection object in notification payload

taylorthurlow opened the pull request:
## What does this pull request do?

ActiveRecord's notification subscription system is leveraged in order to instrument SQL queries. This notification system passes a `payload` hash containing arbitrary notification information about the event that has occurred. `apm-agent-ruby` parses this notification payload and determines the adapter name by looking for the connection object stored in `payload[:connection]`.

This value is _not_ set in any ActiveRecord version before 6.0. In https://github.com/rails/rails/commit/6d1c9a9b23579d13b7cbdd3217ed2cfcc300e06e, it can be seen that the `connection` key is added to the payload, and this change is included in ActiveRecord 6.0 and newer. Earlier versions of ActiveRecord include a `:connection_id` payload key/value pair which includes the ActiveRecord connection object's `#object_id`.

This pull request adds functionality which loads the connection object in the event that `:connection_id` has a value, and `:connection` does not. Because we're loading the object from an `object_id`, proper error handling is included in the event that the object referred to by the `:connection_id` value has been garbage collected (`RangeError` will be raised). The nature of fetching an object by an `object_id` in this way is that we cannot guarantee that the same ID will still match to the object we expect (though we can be reasonably sure). To give ourselves a little more assurance I make sure the resulting object responds to `:adapter_name` as we expect.

It should also be noted that the `:connection_id` key/value pair was removed from the notification payload in https://github.com/rails/rails/commit/d140ab073f296a2a689a4dbd58cf9434c6b2a484#diff-460f4e7973c5dd945c51d24df5b0173961190d3645f4e2585fd3003fa1fc0ff7 (included in ActiveRecord releases 6.1.0 and later), and the included code change is robust enough to handle this key being absent.

## Why is it important?

Currently, if the `payload[:connection]` value is missing (which it will be in AR < 6.0), the fallback adapter value is the result of calling `::ActiveRecord::Base.connection_config[:adapter]`. This is fine if the connection in question uses the base `ActiveRecord::Base` class, but in the apps I'm working on, that isn't a safe assumption. Multiple database connections use multiple base classes which come from various Ruby gems which define their own base connection classes that inherit from `ActiveRecord::Base`.

This change makes things a little more flexible, and we can trust that ActiveRecord will provide us with the correct `object_id` reference to the connection.

One side-effect of this change would be that the aforementioned fallback adapter code path will probably never be reached anymore - The likelihood of the payload missing a value for both `:connection` and `:connection_id` is slim to none. That said, I don't necessarily see the harm in removing it.

## Checklist

- [X] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
- [X] My code follows the style guidelines of this project (See `.rubocop.yml`)
- [X] I have rebased my changes on top of the latest master branch
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing [**unit** tests](https://github.com/elastic/apm-agent-ruby/blob/master/CONTRIBUTING.md#testing) pass locally with my changes
- [ ] ~~I have made corresponding changes to the documentation~~
- [X] I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)
- [ ] ~~I have updated [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)~~
- [ ] ~~Added an API method or config option? Document in which version this will be introduced~~

## Related issues

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-ruby%2Fapm-agent-ruby-mbp%2FPR-953/detail/PR-953/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-ruby%2Fapm-agent-ruby-mbp%2FPR-953/detail/PR-953/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-ruby%2Fapm-agent-ruby-mbp%2FPR-953/detail/PR-953/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-ruby%2Fapm-agent-ruby-mbp%2FPR-953/detail/PR-953/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #953 opened`


* Reason: `The PR is not allowed to run in the CI yet`


* * Start Time: 2021-02-03T01:59:25.872+0000


* Duration: 3 min 48 sec


* Commit: 72467cf8342dedd0e9205188425695f786e397ff




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->







### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-ruby%2Fapm-agent-ruby-mbp%2FPR-953/detail/PR-953/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-ruby/pipelines/apm-agent-ruby-mbp/pipelines/PR-953/runs/1/steps/85/log/?start=0">here</a></li>
<li>Description: <code>githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permission</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-02-03T02:01:45.439Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-ruby.git # timeout=10
[2021-02-03T02:01:45.453Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-02-03T02:01:45.473Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-ruby.git # timeout=10
[2021-02-03T02:01:45.489Z]  > git rev-parse --verify HEAD # timeout=10
[2021-02-03T02:01:45.496Z] No valid HEAD. Skipping the resetting
[2021-02-03T02:01:45.496Z]  > git clean -fdx # timeout=10
[2021-02-03T02:01:45.512Z] Fetching upstream changes from git@github.com:elastic/apm-agent-ruby.git
[2021-02-03T02:01:45.512Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-02-03T02:01:45.516Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/apm-agent-ruby.git +refs/pull/953/head:refs/remotes/origin/PR-953 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2021-02-03T02:01:46.153Z]  > git config core.sparsecheckout # timeout=10
[2021-02-03T02:01:46.172Z]  > git checkout -f 72467cf8342dedd0e9205188425695f786e397ff # timeout=15
[2021-02-03T02:01:46.229Z]  > git remote # timeout=10
[2021-02-03T02:01:46.241Z]  > git config --get remote.origin.url # timeout=10
[2021-02-03T02:01:46.247Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-02-03T02:01:46.251Z]  > git merge 9e49beb4d41538abd25b8b00babf9cbb19989224 # timeout=10
[2021-02-03T02:01:46.264Z]  > git rev-parse HEAD^{commit} # timeout=10
[2021-02-03T02:01:46.274Z]  > git config core.sparsecheckout # timeout=10
[2021-02-03T02:01:46.279Z]  > git checkout -f 72467cf8342dedd0e9205188425695f786e397ff # timeout=15
[2021-02-03T02:01:50.033Z] Commit message: "Load connection object from object_id in AR notification payload parsing"
[2021-02-03T02:01:50.033Z] First time build. Skipping changelog.
[2021-02-03T02:01:50.033Z] Cleaning workspace
[2021-02-03T02:01:50.129Z]  > git --version # timeout=10
[2021-02-03T02:01:50.138Z]  > git --version # 'git version 2.17.1'
[2021-02-03T02:01:50.144Z] fatal: bad object 48a15103e04816e78fb1bdcf6c8cb6ad37e6a8bd
[2021-02-03T02:01:50.038Z]  > git rev-parse --verify HEAD # timeout=10
[2021-02-03T02:01:50.052Z] Resetting working tree
[2021-02-03T02:01:50.052Z]  > git reset --hard # timeout=10
[2021-02-03T02:01:50.083Z]  > git clean -fdx # timeout=10
[2021-02-03T02:01:50.977Z] Masking supported pattern matches of $JOB_GCS_BUCKET or $NOTIFY_TO
[2021-02-03T02:01:51.074Z] Timeout set to expire in 2 hr 0 min
[2021-02-03T02:01:51.105Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2021-02-03T02:01:51.337Z] [INFO] Number of builds to be searched 10
[2021-02-03T02:01:51.513Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2021-02-03T02:01:51.538Z] Running in /var/lib/jenkins/workspace/t-ruby_apm-agent-ruby-mbp_PR-953/src/github.com/elastic/apm-agent-ruby
[2021-02-03T02:01:51.584Z] [INFO] gitCheckout: Checkout SCM PR-953 with some customisation.
[2021-02-03T02:01:51.624Z] [INFO] Override default checkout
[2021-02-03T02:01:51.703Z] Sleeping for 10 sec
[2021-02-03T02:02:01.855Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2021-02-03T02:02:01.955Z] Cloning the remote Git repository
[2021-02-03T02:02:01.973Z] Cloning repository git@github.com:elastic/apm-agent-ruby.git
[2021-02-03T02:02:01.999Z]  > git init /var/lib/jenkins/workspace/t-ruby_apm-agent-ruby-mbp_PR-953/src/github.com/elastic/apm-agent-ruby # timeout=10
[2021-02-03T02:02:02.011Z] Using reference repository: /var/lib/jenkins/.git-references/apm-agent-ruby.git
[2021-02-03T02:02:02.012Z] Fetching upstream changes from git@github.com:elastic/apm-agent-ruby.git
[2021-02-03T02:02:02.012Z]  > git --version # timeout=10
[2021-02-03T02:02:02.017Z]  > git --version # 'git version 2.17.1'
[2021-02-03T02:02:02.017Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-02-03T02:02:02.022Z]  > git fetch --tags --progress -- git@github.com:elastic/apm-agent-ruby.git +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-02-03T02:02:03.168Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-ruby.git # timeout=10
[2021-02-03T02:02:03.173Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2021-02-03T02:02:03.184Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-ruby.git # timeout=10
[2021-02-03T02:02:03.194Z] Fetching upstream changes from git@github.com:elastic/apm-agent-ruby.git
[2021-02-03T02:02:03.194Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2021-02-03T02:02:03.198Z]  > git fetch --tags --progress -- git@github.com:elastic/apm-agent-ruby.git +refs/pull/953/head:refs/remotes/origin/PR-953 +refs/heads/master:refs/remotes/origin/master # timeout=10
[2021-02-03T02:02:03.923Z] Checking out Revision 72467cf8342dedd0e9205188425695f786e397ff (origin/PR-953)
[2021-02-03T02:02:03.973Z] Commit message: "Load connection object from object_id in AR notification payload parsing"
[2021-02-03T02:02:03.997Z]  > git --version # timeout=10
[2021-02-03T02:02:04.005Z]  > git --version # 'git version 2.17.1'
[2021-02-03T02:02:04.013Z] fatal: bad object 48a15103e04816e78fb1bdcf6c8cb6ad37e6a8bd
[2021-02-03T02:02:03.914Z]  > git rev-parse origin/PR-953^{commit} # timeout=10
[2021-02-03T02:02:03.929Z]  > git config core.sparsecheckout # timeout=10
[2021-02-03T02:02:03.932Z]  > git checkout -f 72467cf8342dedd0e9205188425695f786e397ff # timeout=10
[2021-02-03T02:02:05.006Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2021-02-03T02:02:05.617Z] + git fetch https://****:****@github.com/elastic/apm-agent-ruby.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2021-02-03T02:02:06.265Z] Running in /var/lib/jenkins/workspace/t-ruby_apm-agent-ruby-mbp_PR-953/src/github.com/elastic/apm-agent-ruby/.git
[2021-02-03T02:02:06.310Z] Archiving artifacts
[2021-02-03T02:02:07.050Z] + git rev-parse HEAD
[2021-02-03T02:02:07.492Z] + git rev-parse HEAD
[2021-02-03T02:02:07.803Z] + git rev-parse origin/pr/953
[2021-02-03T02:02:07.856Z] [INFO] githubEnv: Found Git Build Cause: pr
[2021-02-03T02:02:08.107Z] Masking supported pattern matches of $GITHUB_TOKEN
[2021-02-03T02:02:09.029Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/apm-agent-ruby/pulls/953/reviews return 0 elements
[2021-02-03T02:02:09.095Z] [INFO] githubPrCheckApproved: Title: Fix missing ActiveRecord connection object in notification payload - User: taylorthurlow - Author Association: FIRST_TIME_CONTRIBUTOR
[2021-02-03T02:02:09.450Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2021-02-03T02:02:09.450Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2021-02-03T02:02:09.540Z] [INFO] Let's stop build #1. The PR is not allowed to run in the CI yet
[2021-02-03T02:02:09.573Z] Sleeping for 5 sec
[2021-02-03T02:02:10.653Z] Stage "Sanity checks" skipped due to earlier failure(s)
[2021-02-03T02:02:10.747Z] Stage "Tests" skipped due to earlier failure(s)
[2021-02-03T02:02:10.873Z] Stage "Tests" skipped due to earlier failure(s)
[2021-02-03T02:02:10.884Z] Stage "Master Tests frameworks" skipped due to earlier failure(s)
[2021-02-03T02:02:10.895Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2021-02-03T02:02:11.005Z] Failed in branch Tests
[2021-02-03T02:02:11.014Z] Failed in branch Master Tests frameworks
[2021-02-03T02:02:11.022Z] Failed in branch Integration Tests
[2021-02-03T02:02:11.217Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-02-03T02:02:11.260Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-02-03T02:02:11.358Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2021-02-03T02:02:11.524Z] Stage "Release" skipped due to earlier failure(s)
[2021-02-03T02:02:11.565Z] Stage "Release" skipped due to earlier failure(s)
[2021-02-03T02:02:11.651Z] Stage "Release" skipped due to earlier failure(s)
[2021-02-03T02:02:11.737Z] Stage "Release" skipped due to earlier failure(s)
[2021-02-03T02:02:12.525Z] Running on Jenkins in /var/lib/jenkins/workspace/t-ruby_apm-agent-ruby-mbp_PR-953
[2021-02-03T02:02:12.663Z] [INFO] getVaultSecret: Getting secrets
[2021-02-03T02:02:12.837Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-02-03T02:02:13.770Z] + chmod 755 generate-build-data.sh
[2021-02-03T02:02:13.770Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-ruby/apm-agent-ruby-mbp/PR-953/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-ruby/apm-agent-ruby-mbp/PR-953/runs/1 ABORTED 167622
[2021-02-03T02:02:14.020Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-ruby/apm-agent-ruby-mbp/PR-953/runs/1/steps/?limit=10000 -o steps-info.json
[2021-02-03T02:02:14.931Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-ruby/apm-agent-ruby-mbp/PR-953/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-02-03T02:02:14.931Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2021-02-03T02:02:15.842Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

estolfo commented on spec/elastic_apm/normalizers/active_record_spec.rb:
```diff
@@ -90,6 +90,58 @@ def normalize_payload(payload)
+            _name, _type, subtype, = normalize_payload(
+              sql: sql,
+              connection: double(adapter_name: 'MySQL'),
+              connection_id: 'wrong object'.object_id
```
It might make more sense for this to be `connection_id: double(adapter_name: 'wrong_db').object_id`, to be sure that we could successfully get an adapter name from the `connection_id:` field, if there were a bug in the code.

taylorthurlow replied:
Seems like a good change to me, thanks.

estolfo commented on spec/elastic_apm/normalizers/active_record_spec.rb:
```diff
@@ -90,6 +90,58 @@ def normalize_payload(payload)
+              sql: sql,
+              connection_id: double(adapter_name: 'Postgres').object_id
+            )
+            expect(subtype).to eq 'postgres'
```
What is the reason for testing this twice with a different `adapter_name`?

taylorthurlow replied:
Pretty much only because the comparable pre-existing test for using the `:connection` value checked both. The only assumption I might be able to make about the original tester's intent was to check that the adapter name gets downcased properly?

See this existing test on `master`: https://github.com/elastic/apm-agent-ruby/blob/master/spec/elastic_apm/normalizers/active_record_spec.rb#L77-L92

estolfo replied:
Thanks for the explanation. I don't know why exactly we have more than one check for downcasing the adapter name. I guess we can just skip doing a second check in your tests.

estolfo reviewed (changes requested):
Thank you so much for the thorough explanation and fantastic code changes + tests! I had a few suggestions/comments.

taylorthurlow commented on lib/elastic_apm/normalizers/rails/active_record.rb:
```diff
@@ -57,6 +57,18 @@ def normalize(_transaction, _name, payload)
 
         def subtype_for(payload)
+          if payload[:connection_id]
+            payload[:connection] ||= begin
```
I went back and forth on this idea. It seems like it actually is our own object, we subscribed to notifications from ActiveRecord and they sent us a copy. If I receive a magazine in the mail every month because I asked for it, it's my magazine. We read it and throw it away, don't we?

Maybe my analogy doesn't fit, but it doesn't make any difference to me in the end. I think I went this route because it required less code overall. I'll refactor this method to skip the mutation. I also assume, based on your suggestion, that you think we should remove the `::ActiveRecord::Base.connection_config[:adapter]` check? I mentioned that it was pretty much redundant, as I don't expect there to be a case where that code is reached anyways.

taylorthurlow commented:
Thanks for reviewing so quickly! Left my feedback where necessary, I'll make some of the changes now.

taylorthurlow commented:
Ok so I've pushed up another commit which addresses the payload mutation discussion. I removed the code path which uses `::ActiveRecord::Base` as a fallback, and the tests associated with it. I believe I found a problem with the test suite, though - I removed this bit of code now that the system under test doesn't refer to the class anymore:

https://github.com/elastic/apm-agent-ruby/blob/9e49beb4d41538abd25b8b00babf9cbb19989224/spec/elastic_apm/normalizers/active_record_spec.rb#L36-L40

This had the unintended impact of breaking the spec located at `spec/elastic_apm/spies/shoryuken_spec.rb`:

```
$> be rspec
[INFO] Skipping Sinatra.start spec
Run options: exclude {:type=>"json_schema"}

Randomized with seed 54416
...........................................................................................................................................................FF.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................*..........................................................................................................

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) ElasticAPM::Metrics thread safety stress test handles multiple threads reporting and collecting at the same time
     # Temporarily skipped with xcontext
     # ./spec/elastic_apm/metrics_spec.rb:118


Failures:

  1) Spy: Shoryuken reports errors
     Failure/Error:
       allow(::ActiveRecord::Base)
         .to receive(:clear_active_connections!)

     NameError:
       uninitialized constant ActiveRecord
     # ./spec/elastic_apm/spies/shoryuken_spec.rb:68:in `block (2 levels) in <module:ElasticAPM>'
     # ./vendor/ruby/2.7.0/gems/webmock-3.11.2/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

  2) Spy: Shoryuken instruments jobs
     Failure/Error:
       allow(::ActiveRecord::Base)
         .to receive(:clear_active_connections!)

     NameError:
       uninitialized constant ActiveRecord
     # ./spec/elastic_apm/spies/shoryuken_spec.rb:68:in `block (2 levels) in <module:ElasticAPM>'
     # ./vendor/ruby/2.7.0/gems/webmock-3.11.2/lib/webmock/rspec.rb:37:in `block (2 levels) in <top (required)>'

Finished in 11.84 seconds (files took 1.75 seconds to load)
767 examples, 2 failures, 1 pending

Failed examples:

rspec ./spec/elastic_apm/spies/shoryuken_spec.rb:106 # Spy: Shoryuken reports errors
rspec ./spec/elastic_apm/spies/shoryuken_spec.rb:89 # Spy: Shoryuken instruments jobs

Randomized with seed 54416
```

The tests do run in random order so I imagine this could be an intermittent issue. I added the snippet above into this newly-failing test and everything is passing again.

Regarding the changes that I made to avoid the payload mutation, let me know if you think it's an appropriate solution. I think I arrived at an implementation that is reasonably easy to grep.

estolfo commented:
Sorry about that `ActiveRecord::Base` trouble. I looked into it and it seems that we were defining an `ActiveRecord` class and `ActiveRecord::Base` class outside of a `before` block in the `active_record_spec`, which meant that the classes were defined when the file was loaded. Regardless of what order the specs ran in, the classes would have been defined.

Anyway, what you've done-- putting the definition in a `before` block of the Shoryuken spec is a good solution so those classes are defined when the spec is run.

estolfo commented on lib/elastic_apm/normalizers/rails/active_record.rb:
```diff
@@ -57,10 +57,17 @@ def normalize(_transaction, _name, payload)
+            begin
+              loaded_object = ObjectSpace._id2ref(payload[:connection_id])
+              return cached_adapter_name(loaded_object.adapter_name) if loaded_object.respond_to?(:adapter_name)
+            rescue RangeError # if connection object has somehow been garbage collected
```
We return `unknown` if we run into any trouble getting the connection object or its adapter name. So I think we should also return `unknown` in the case we can't get the connection object because it has been garbage collected.

estolfo replied:
I think you also removed a test for this code path. Was that intentional?
https://github.com/elastic/apm-agent-ruby/pull/953/commits/c776010ff157e203cf5c128055366e9f8c43c525#diff-54b1044cee5cfa350d537561412b450f0858a900078239714e4f503306491251L108

taylorthurlow replied:
> We return `unknown` if we run into any trouble getting the connection object or its adapter name. So I think we should also return `unknown` in the case we can't get the connection object because it has been garbage collected.

Unless I misunderstand the code path you're referring to, I believe this does what you want. the final non-early-return case in `#subtype_for` calls `cached_adapter_name(nil)` which will return the adapter name in the `UNKNOWN` constant. The only cases in which we won't reach this final return value are:

- If we have a `payload[:connection]` value and we grab it's adapter name
- If we have a `payload[:connection_id]` value, it loads an object correctly, _and_ that object responds to `:adapter_name`

Otherwise that final code path should be reached and we'll get `unknown` as the adapter name.

> I think you also removed a test for this code path. Was that intentional?
> [c776010#diff-54b1044cee5cfa350d537561412b450f0858a900078239714e4f503306491251L108](https://github.com/elastic/apm-agent-ruby/commit/c776010ff157e203cf5c128055366e9f8c43c525#diff-54b1044cee5cfa350d537561412b450f0858a900078239714e4f503306491251L108)

This one is sort of an oversight on my part. I'm still going to "remove" this test but I'll be replacing it with one or two that properly test the fallback to the `unknown` adapter.

taylorthurlow replied:
Ok I pushed up the new tests. I don't see a way to effectively test the code path where the object_id lookup raises the `RangeError` due to the object having been garbage collected.

We'd somehow have to manually garbage collect an object, which, even if I knew how to do, I don't think would be a good idea. Do you have any suggestions, or are you ok with that code path remaining untested?

estolfo commented:
jenkins, please run the tests

estolfo reviewed (commented):
Thanks a lot for addressing my comments!

I think there might be an issue with using `ObjectSpace._id2ref` and I apologize for not realizing this earlier.
I was reading about the method and how it had changed in ruby 2.7 and I found [this issue](https://bugs.ruby-lang.org/issues/15711) in the Ruby tracker.
It says that the method will be deprecated soon. So this is a good solution now but unfortunately it’s not a long-term one.

Secondly, I read in Charles Nutter’s comments that JRuby only supports `_id2ref` with a command line flag. We officially support JRuby and we unfortunately can’t require our users to use a particular command line flag for the agent to work properly. I ran the tests on CI, and as expected, the test fails on JRuby. The method raises a RangeError if the command line flags are not used.
Here is the behavior in irb:

```
jruby-9.2.8.0 :002 > "my string".object_id
 => 2002 
jruby-9.2.8.0 :003 > ObjectSpace._id2ref(2002)
(irb):3: warning: ObjectSpace is disabled; _id2ref only supports immediates, pass -X+O to enable
Traceback (most recent call last):
        7: from /Users/emily/.rvm/rubies/jruby-9.2.8.0/bin/irb:13:in `<main>'
        6: from org/jruby/RubyKernel.java:1193:in `catch'
        5: from org/jruby/RubyKernel.java:1193:in `catch'
        4: from org/jruby/RubyKernel.java:1425:in `loop'
        3: from org/jruby/RubyKernel.java:1061:in `eval'
        2: from (irb):3:in `evaluate'
        1: from org/jruby/RubyObjectSpace.java:114:in `_id2ref'
RangeError (0x00000000000007d2 is not id value)
```


Considering `_id2ref` will eventually be deprecated and that we can’t require JRuby users to pass certain command line args in order for this code to work, it doesn’t seem to make sense to use this method. I’m open to ideas though— you could write the code such that it fails back to our previous behavior using `::ActiveRecord::Base.connection_config[:adapter]` for JRuby or when a RangeError is raised.

And finally, regarding the test for a RangeError, you can use `_id2ref` with 4 and it’ll always throw a RangeError.

Let me know what you think!

estolfo commented:
Hi @taylorthurlow, do you have any thoughts on my last comment?

taylorthurlow commented:
Hey, sorry for the delay.

> I think there might be an issue with using `ObjectSpace._id2ref` and I apologize for not realizing this earlier. I was reading about the method and how it had changed in ruby 2.7 and I found [this issue](https://bugs.ruby-lang.org/issues/15711) in the Ruby tracker. It says that the method will be deprecated soon. So this is a good solution now but unfortunately it’s not a long-term one.

Interesting discussion on this issue, I wasn't aware of the shortcomings of the method.

> Secondly, I read in Charles Nutter’s comments that JRuby only supports `_id2ref` with a command line flag. We officially support JRuby and we unfortunately can’t require our users to use a particular command line flag for the agent to work properly. I ran the tests on CI, and as expected, the test fails on JRuby. The method raises a RangeError if the command line flags are not used.

I hadn't thought about this, so good thing you brought it up!

> regarding the test for a RangeError, you can use `_id2ref` with 4 and it’ll always throw a RangeError.

Nice tip! Handy.

> Considering `_id2ref` will eventually be deprecated and that we can’t require JRuby users to pass certain command line args in order for this code to work, it doesn’t seem to make sense to use this method. I’m open to ideas though— you could write the code such that it fails back to our previous behavior using `::ActiveRecord::Base.connection_config[:adapter]` for JRuby or when a RangeError is raised.

So given this new information I see two ways forward.

To fix the issue I could do a check that we are using MRI and only attempt the use of `id2ref` if the method is defined. I could also optionally add a check for an ActiveRecord version `< 6`, while not strictly necessary, it would probably avoid potential deprecation warning messages in the future. As you mentioned the `ActiveRecord::Base` fallback would be added back, seeing as we can't assume this new functionality works in all cases.

Alternatively, maybe the best choice is to just continue with the subpar details in older ActiveRecord versions.

I'm OK either way, so let me know if you think the extra checks are worth it, and I can fix things up. Thanks for checking in again!

estolfo commented:
Hi @taylorthurlow, thanks for your response. Most of our Rails users are on MRI so I think that would be a reasonable check. Then, if `_id2ref` is not available, we can fall back to using `ActiveRecord::Base`. I don't think it's necessary to check the `ActiveRecord` version as the deprecation doesn't have a target date so it's probably not worth worrying about the deprecation warning for now. Thank you for your effort on this!

taylorthurlow commented:
jenkins, please run the tests

taylorthurlow commented:
Sorry for the delay @estolfo. I just pushed up changes which should achieve what we were looking for. I added a check for the current Ruby interpreter and added the `::ActiveRecord::Base` check back in. In order to properly test this stuff I added a little skip clause to the relevant tests which should mark the tests as pending/skipped for non-MRI environments. The skip message explains that they are MRI-only. Let me know if that solution is adequate or if you'd prefer a solution that pretends those tests don't exist rather than marking them as skipped.

estolfo commented:
jenkins, please run the tests

estolfo commented:
jenkins, please run the tests

estolfo commented:
Thank you so much for your work on this, @taylorthurlow!

taylorthurlow commented:
No problem, glad to contribute. Thanks for your detailed feedback.
